### ISSUE TITLE
ClientError: SignatureDoesNotMatch

### ISSUE THREAD

**Issue by cmuell89** [jschneier/django-storages#386]
I have a DRF application that uses django-storages to store media files in S3. In development, everything works smoothly and files are uploaded to the designated S3 bucket. However, after launching the application via Heroku, I repeatedly get the following error:

**Error**
botocore.exceptions.ClientError: An error occurred (SignatureDoesNotMatch) when calling the PutObject operation: The request signature we calculated does not match the signature you provided. Check your key and signing method.

**Trace**
```
2017-08-25T22:56:24.412631+00:00 app[web.1]: <rest_framework.request.Request object at 0x7f8561d1d748>
2017-08-25T22:56:24.802971+00:00 app[web.1]: Internal Server Error: /api/v1/opportunities/1/media/
2017-08-25T22:56:24.802984+00:00 app[web.1]: Traceback (most recent call last):
2017-08-25T22:56:24.802985+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
2017-08-25T22:56:24.802986+00:00 app[web.1]:     response = get_response(request)
2017-08-25T22:56:24.802990+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
2017-08-25T22:56:24.802991+00:00 app[web.1]:     response = self.process_exception_by_middleware(e, request)
2017-08-25T22:56:24.802992+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
2017-08-25T22:56:24.802994+00:00 app[web.1]:     response = wrapped_callback(request, *callback_args, **callback_kwargs)
2017-08-25T22:56:24.802994+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
2017-08-25T22:56:24.802995+00:00 app[web.1]:     return view_func(*args, **kwargs)
2017-08-25T22:56:24.802995+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
2017-08-25T22:56:24.802996+00:00 app[web.1]:     return self.dispatch(request, *args, **kwargs)
2017-08-25T22:56:24.802997+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/rest_framework/views.py", line 489, in dispatch
2017-08-25T22:56:24.802997+00:00 app[web.1]:     response = self.handle_exception(exc)
2017-08-25T22:56:24.802998+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/rest_framework/views.py", line 449, in handle_exception
2017-08-25T22:56:24.802999+00:00 app[web.1]:     self.raise_uncaught_exception(exc)
2017-08-25T22:56:24.802999+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/rest_framework/views.py", line 486, in dispatch
2017-08-25T22:56:24.803000+00:00 app[web.1]:     response = handler(request, *args, **kwargs)
2017-08-25T22:56:24.803001+00:00 app[web.1]:   File "/app/verigive/api/views.py", line 777, in put
2017-08-25T22:56:24.803002+00:00 app[web.1]:     opportunity.save()
2017-08-25T22:56:24.803002+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/base.py", line 806, in save
2017-08-25T22:56:24.803003+00:00 app[web.1]:     force_update=force_update, update_fields=update_fields)
2017-08-25T22:56:24.803003+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/base.py", line 836, in save_base
2017-08-25T22:56:24.803004+00:00 app[web.1]:     updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
2017-08-25T22:56:24.803005+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/base.py", line 900, in _save_table
2017-08-25T22:56:24.803005+00:00 app[web.1]:     for f in non_pks]
2017-08-25T22:56:24.803006+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/base.py", line 900, in <listcomp>
2017-08-25T22:56:24.803007+00:00 app[web.1]:     for f in non_pks]
2017-08-25T22:56:24.803007+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/fields/files.py", line 296, in pre_save
2017-08-25T22:56:24.803008+00:00 app[web.1]:     file.save(file.name, file.file, save=False)
2017-08-25T22:56:24.803009+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/db/models/fields/files.py", line 94, in save
2017-08-25T22:56:24.803009+00:00 app[web.1]:     self.name = self.storage.save(name, content, max_length=self.field.max_length)
2017-08-25T22:56:24.803010+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/django/core/files/storage.py", line 54, in save
2017-08-25T22:56:24.803010+00:00 app[web.1]:     return self._save(name, content)
2017-08-25T22:56:24.803011+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/storages/backends/s3boto3.py", line 452, in _save
2017-08-25T22:56:24.803012+00:00 app[web.1]:     self._save_content(obj, content, parameters=parameters)
2017-08-25T22:56:24.803012+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/storages/backends/s3boto3.py", line 467, in _save_content
2017-08-25T22:56:24.803013+00:00 app[web.1]:     obj.upload_fileobj(content, ExtraArgs=put_parameters)
2017-08-25T22:56:24.803013+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/boto3/s3/inject.py", line 509, in object_upload_fileobj
2017-08-25T22:56:24.803014+00:00 app[web.1]:     ExtraArgs=ExtraArgs, Callback=Callback, Config=Config)
2017-08-25T22:56:24.803014+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/boto3/s3/inject.py", line 427, in upload_fileobj
2017-08-25T22:56:24.803015+00:00 app[web.1]:     return future.result()
2017-08-25T22:56:24.803016+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/s3transfer/futures.py", line 73, in result
2017-08-25T22:56:24.803016+00:00 app[web.1]:     return self._coordinator.result()
2017-08-25T22:56:24.803017+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/s3transfer/futures.py", line 233, in result
2017-08-25T22:56:24.803017+00:00 app[web.1]:     raise self._exception
2017-08-25T22:56:24.803018+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/s3transfer/tasks.py", line 126, in __call__
2017-08-25T22:56:24.803019+00:00 app[web.1]:     return self._execute_main(kwargs)
2017-08-25T22:56:24.803020+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/s3transfer/tasks.py", line 150, in _execute_main
2017-08-25T22:56:24.803020+00:00 app[web.1]:     return_value = self._main(**kwargs)
2017-08-25T22:56:24.803021+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/s3transfer/upload.py", line 679, in _main
2017-08-25T22:56:24.803021+00:00 app[web.1]:     client.put_object(Bucket=bucket, Key=key, Body=body, **extra_args)
2017-08-25T22:56:24.803022+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/botocore/client.py", line 310, in _api_call
2017-08-25T22:56:24.803023+00:00 app[web.1]:     return self._make_api_call(operation_name, kwargs)
2017-08-25T22:56:24.803030+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/botocore/client.py", line 599, in _make_api_call
2017-08-25T22:56:24.803031+00:00 app[web.1]:     raise error_class(parsed_response, operation_name)
2017-08-25T22:56:24.803033+00:00 app[web.1]: botocore.exceptions.ClientError: An error occurred (SignatureDoesNotMatch) when calling the PutObject operation: The request signature we calculated does not match the signature you provided. Check your key and signing method.
```

**Associated view**
```
class OpportunityMediaUpdate(generics.GenericAPIView):
    """
    Class-based view supporting put/patch request to upload media image for the specified Opportunity, if permitted.

    Attributes:
        authentication_class (tuple): Uses JWT authentication.
        permission_classes (tuple): Requires authentication.
        serializer_class (cls): Serializer class providing serialized Opportunity.
    """
    authentication_classes = (JSONWebTokenAuthentication,)
    permission_classes = (IsAuthenticated,)
    serializer_class = OpportunitySerializer
    parser_classes = (MultiPartParser,)

    def put(self, request, pk):
        opportunity = get_object_or_404(Opportunity, id=pk)
        serializer = self.get_serializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        data = serializer.validated_data
        header_image = data['header_image']
        print(str(request))
        opportunity.header_image = header_image
        opportunity.save()
        serializer = OpportunitySerializer(opportunity)
        return Response(serializer.data)
```

**Associated Model**
Notes:
- upload_to is a simple function that returns back a string for the directory structure in S3.
- header_image field is the only model field used to reference media.
```
class Opportunity(CreateModified):
    file_size_limiter = partial(file_size, megabytes=1)
    media_uploader = partial(upload_to, model_name='opportunities')
    ACTIVE = 'active'
    INACTIVE = 'inactive'
    DRAFT = 'draft'
    IN_REVIEW = 'in-review'
    STATUS_CHOICES = [(DRAFT, 'Draft'), (ACTIVE, 'Active'),
                      (INACTIVE, 'Inactive'), (IN_REVIEW, 'In-Review')]

    APPROVED = 'approved'
    UNAPPROVED = 'unapproved'
    HOST_APPROVAL_CHOICES = [(APPROVED, 'Approved'),
                             (UNAPPROVED, "Unapproved")]

    name = models.CharField(max_length=150, blank=True)
    blurb = models.CharField(max_length=100, blank=True)
    description = models.TextField(blank=True)
    header_image = models.ImageField(
        _('image'), blank=True, null=True, upload_to=media_uploader, validators=[file_size_limiter])
    host_approval = models.CharField(
        max_length=50, choices=HOST_APPROVAL_CHOICES, default=UNAPPROVED)
    status = models.CharField(max_length=50, choices=STATUS_CHOICES,
                              default=DRAFT)
    organization = models.ForeignKey(
        'BeneficiaryOrganization', related_name='opportunities', on_delete=models.CASCADE, blank=True)
    tags = models.ManyToManyField('Interest', blank=True)
    expiration_date = models.DateTimeField(null=True, blank=True)
    target_amount = models.IntegerField(default=None, null=True, blank=True)
    suggested_amount = models.IntegerField(default=None, null=True, blank=True)
```

**Applicable Settings**
```
"""
STATIC/MEDIA FILES
"""
# https://www.caktusgroup.com/blog/2014/11/10/Using-Amazon-S3-to-store-your-Django-sites-static-and-media-files/
AWS_ACCESS_KEY_ID = os.environ['AWS_ACCESS_KEY_ID']
AWS_SECRET_ACCESS_KEY = os.environ['AWS_SECRET_ACCESS_KEY']
AWS_STORAGE_BUCKET_NAME = os.environ['AWS_STORAGE_BUCKET_NAME']
AWS_S3_CUSTOM_DOMAIN = '{}.s3.amazonaws.com'.format(AWS_STORAGE_BUCKET_NAME)
AWS_S3_REGION_NAME = os.environ['AWS_S3_REGION_NAME']
AWS_S3_ENCRYPTION = os.environ['AWS_S3_ENCRYPTION']
```

**Environment Variables**
Sensitive info hidden.
```
AWS_ACCESS_KEY_ID=***********
AWS_SECRET_ACCESS_KEY=***********
AWS_STORAGE_BUCKET_NAME=media
AWS_S3_REGION_NAME=us-west-2
AWS_S3_ENCRYPTION=True
```

**Pertinent Packages**
```
boto3==1.4.4
botocore==1.5.79
django-storages==1.6.5
```

I am not sure why in a development setting, things function smoothly, but not in a production setting. Any hints, tips or advice would be great.

**Comment by jschneier** [comment]
https://github.com/jschneier/django-storages/issues/343#issuecomment-310771623

I think this is the issue. They switched the default signature version on us.

**Comment by jschneier** [comment]
Can you confirm and I guess I'll ship a version that defaults us to `s3` and then issue a `2.0` sometime in the next few weeks that will default to `s3v4`.

**Comment by cmuell89** [comment]
@jschneier I decided to redo my IAM access token after following this issue: https://github.com/aws/aws-cli/issues/602

I made a secret key without any slashes or special characters. 

I also removed from settings.py:
AWS_S3_REGION_NAME=us-west-2
AWS_S3_ENCRYPTION=True

It is now working. I cannot say what did the trick entirely. Perhaps my access key was invalid or was not being passed along correctly.
